EventListener/DncGetSubscriber.php

```diff
@@ -0,0 +1,183 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: nbush
+ * Date: 3/5/19
+ * Time: 10:39 AM.
+ */
+
+namespace MauticPlugin\MauticDoNotContactExtrasBundle\EventListener;
+
+use Mautic\CoreBundle\EventListener\CommonSubscriber;
+use Mautic\LeadBundle\Entity\DoNotContact;
+use Mautic\LeadBundle\Entity\Lead;
+use Mautic\LeadBundle\Event\LeadDNCGetEntitiesEvent;
+use Mautic\LeadBundle\LeadEvents;
+
+//use Mautic\LeadBundle\Event\LeadDNCGetCountEvent;
+//use Mautic\LeadBundle\Event\LeadDNCGetEntitiesEvent;
+//use Mautic\LeadBundle\Event\LeadDNCGetListEvent;
+
+/**
+ * Class DncGetSubscriber.
+ */
+class DncGetSubscriber extends CommonSubscriber
+{
+    /**
+     * {@inherit_doc}.
+     *
+     * @return array
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            'mautic.dnc_get_count'    => ['onGetDncCount', 0],
+            'mautic.dnc_get_entities' => ['onGetDncEntities', 0],
+            'mautic.dnc_get_list'     => ['onGetDncList', 0],
+        ];
+    }
+
+    /**
+     * @return \Doctrine\ORM\EntityRepository|\MauticPlugin\MauticDoNotContactExtrasBundle\Entity\DncListItemRepository
+     */
+    protected function getDncItemRepository()
+    {
+        return $this->em->getRepository('MauticDoNotContactExtrasBundle:DncListItem');
+    }
+
+    /**
+     * Returns an array of lead field values to look for in the DNC list based on channel.
+     *
+     * @param Lead   $lead
+     * @param string $channel
+     *
+     * @return mixed
+     */
+    protected function getChannelFieldValues(Lead $lead, $channel)
+    {
+        $channelFieldValues = [];
+
+        switch ($channel) {
+            case 'email':
+                if ($lead->getEmail()) {
+                    $channelFieldValues[] = $lead->getEmail();
+                }
+                break;
+
+            case 'sms':
+            case 'phone':
+            default:
+                if ($lead->getPhone()) {
+                    $channelFieldValues[] = $lead->getPhone();
+                }
+                if ($lead->getMobile()) {
+                    $channelFieldValues[] = $lead->getMobile();
+                }
+        }
+
+        return $channelFieldValues;
+    }
+
+    /**
+     * @param \Mautic\LeadBundle\Event\LeadDNCGetEntitiesEvent $event
+     */
+    public function onGetDncEntities(LeadDNCGetEntitiesEvent $event)
+    {
+        $lead    =  $event->getLead();
+        $channel = $event->getChannel();
+
+        $dncSearch = $this->getChannelFieldValues($lead, $channel);
+        if (empty($dncSearch)) {
+            return;
+        }
+
+        $dbalQb = $this->em->getConnection()->createQueryBuilder()
+            ->select('dnc.reason, dnc.date_added')
+            ->from('dnc_extras_list_items', 'dnc')
+            ->where('dnc.channel = :channel')
+            ->setParameter('channel', $channel);
```

Comment:
Should we handle in case channel is null or empty? I know we say WE will always pass a value, but...